src/vmm/src/devices/virtio/iovec.rs

```diff
@@ -251,16 +215,69 @@ impl IoVecBufferMut {
     /// # Returns
     ///
     /// The number of bytes written (if any)
-    pub fn write_at(&mut self, buf: &[u8], offset: usize) -> Option<usize> {
+    pub fn write_at(&mut self, mut buf: &[u8], offset: usize) -> Option<usize> {
         if offset < self.len() {
             // Make sure we only write up to the end of the `IoVecBufferMut`.
             let size = buf.len().min(self.len() - offset);
-            Some(self.write_subregion(&buf[..size], offset))
+            self.write_volatile_at(&mut buf, offset, size).ok()
         } else {
             // We cannot write past the end of the `IoVecBufferMut`.
             None
         }
     }
+
+    /// Writes up to `limit` bytes into the `IoVecBuffer` starting at the given offset.
+    ///
+    /// This will try to write to the given [`WriteVolatile`].
+    pub fn write_volatile_at<W: ReadVolatile>(
+        &mut self,
+        src: &mut W,
+        mut offset: usize,
+        mut limit: usize,
+    ) -> Result<usize, VolatileMemoryError> {
+        let mut total_bytes_read = 0;
+
+        for iov in &self.vecs {
+            if limit == 0 {
+                break;
+            }
+
+            if offset >= iov.iov_len {
+                offset -= iov.iov_len;
+                continue;
+            }
+
+            let mut slice =
+                // SAFETY: the constructor IoVecBufferMut::from_descriptor_chain ensures that
+                // all iovecs contained point towards valid ranges of guest memory
+                unsafe { VolatileSlice::new(iov.iov_base.cast(), iov.iov_len).offset(offset)? };
```

Comment:
I _think_ we should create this with a bitmap, no?

Reply:
In this case, no, because it would create a new bitmap that is completely detached from the bitmap object associated with the GuestMemoryMmap into which these pointers lead. We already mark the entire IoVecBufferMut region as dirty in `from_tx_descriptor_chain`, before we "forget" about the bitmap by only storing pointers.